### PR TITLE
Ractive components

### PR DESCRIPTION
**Pull request by harrislapiroff**
This branch includes:

* Namespacing all Ractive components into `window.components`.
* Isolating all Ractive component javascript files into their own closures. (Exposed through the namespace above.)
* Reorganized javascript directory structure.
* Including [underscore](http://underscorejs.org/) sitewide.
* Introducing a `Stepbar` component that is representative of what a small atomic component can be.

Documentation of individual components to follow.

### CODE REVIEW

**Comment by harrislapiroff** [comment]
This branch does not change any functionality, hopefully.

**Comment by melinath** [comment]
I am so excited to review this!

File: brambling/templates/brambling/event/order/attendees.html [hunk 1]
```diff
@@ -9,46 +9,56 @@
...
-					name: "{{ step.name }}",
-					url: "{% url step.view_name organization_slug=event.organization.slug event_slug=event.slug %}",
-					is_completed: {% if step.is_completed %}true{% else %}false{% endif %},
-					is_accessible: {% if step.is_accessible %}true{% else %}false{% endif %}
-				}{% if not forloop.last %},{% endif %}
-			{% endfor %}
-		]);
+		var stepbar = components.Stepbar.extend({
+				template: '#StepbarRactive',
+				data: function () {
```
Comment (melinath) [inline brambling/templates/brambling/event/order/attendees.html]: Why is this a `function()`?

File: brambling/templates/brambling/event/order/attendees.html [hunk 2]
```diff
@@ -9,46 +9,56 @@
...
+									key: "{{ step.slug }}",
+									name: "{{ step.name }}",
+									url: "{% url step.view_name organization_slug=event.organization.slug event_slug=event.slug %}",
+									is_completed: {% if step.is_completed %}true{% else %}false{% endif %},
+									is_accessible: {% if step.is_accessible %}true{% else %}false{% endif %},
+									is_active: false
+								}{% if not forloop.last %},{% endif %}
+							{% endfor %}
+						],
+						classes: "{% if order.has_cart %}{% else %}stepbar-hug-masthead{% endif %}"
```
Comment (melinath) [inline brambling/templates/brambling/event/order/attendees.html]: why not just `{% if not order.has_cart %}`

File: brambling/static/brambling/javascript/components/register.ractive.js [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                            }
+                        } else {
+                            thisObj.set('order.' + key, value);
+                        }
+                    });
+                    thisObj.countdown();
+                }
+            });
+        },
+        countdown: function() {
```
Comment (melinath) [inline brambling/static/brambling/javascript/components/register.ractive.js]: Should countdown maybe also be a separate component?

Comment (harrislapiroff) [inline brambling/static/brambling/javascript/components/register.ractive.js]: definitely :D

File: brambling/templates/brambling/event/order/attendees.html (hunk 2 above)
Comment (harrislapiroff) [inline brambling/templates/brambling/event/order/attendees.html]: oops

File: brambling/templates/brambling/event/order/attendees.html (hunk 1 above)
Comment (harrislapiroff) [inline brambling/templates/brambling/event/order/attendees.html]: Ractive issues a warning if you assign arrays and objects as data. It's probably not necessary that this be a function, but I thought I'd do what it told me.

Comment (melinath) [inline brambling/templates/brambling/event/order/attendees.html]: Huh. On components specifically? I don't think I've seen that... But okay, w/e. /shrug

File: brambling/static/brambling/javascript/components/register.ractive.js [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                        thisObj.splice('order.bought_items', index, 1);
+                    }
+                    thisObj.loadItems();
+                    thisObj.loadOrder();
+                }
+            });
+        },
+        oninit: function (options) {
+            var thisObj = this;
+            thisObj.observe("discountCode", function (newValue, oldValue) {
```
Comment (melinath) [inline brambling/static/brambling/javascript/components/register.ractive.js]: All this component talk has me wondering if this (discount handling) should be a separate component as well?

**Comment by melinath** [comment]
Do you want to do more component splitting as part of this branch, or would you prefer to have it merged first?

**Comment by harrislapiroff** [comment]
I think I'd prefer to have it merged first, unless you think it'd be better to do more on this branch. The stepbar was really mostly, like, a proof of concept.

**Comment by harrislapiroff** [comment]
I can fix the little things in a bit though.

**Comment by melinath** [comment]
the ractive-shop is... even more broken on this branch. Though it doesn't look like it's the branch's fault...

**Comment by melinath** [comment]
Yeah, broken on master too.

**Comment by melinath** [comment]
I'm down with you merging after the little fixes.

**Pull request merged**